include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -150,7 +161,7 @@ typename MatType::elem_type NSGA2::Optimize(
 
     // Perform crowding distance assignment.
     crowdingDistance.resize(population.size());
-
+    std::fill(crowdingDistance.begin(), crowdingDistance.end(), 0.f);
```

Comment:
I guess before the reset to zero, we got lucky, because in the `CrowdingDistanceAssignment` method we use `front[i + 1]` which should be undefined:

```
crowdingDistance[front[i]] += (crowdingDistance[front[i - 1]] -
            crowdingDistance[front[i + 1]])
```

Reply:
Ahh, nevermind we set it to zero before as well.

Reply:
@zoq  Regarding this, I think this implementation is wrong. Here's an excerpt from the paper

![crowdingdistance](https://user-images.githubusercontent.com/45007169/110422835-19d1ec80-80c6-11eb-984a-702a6954aa76.png)

Notice the equation should've been 
```c++
crowdingDistance[front[i]] += (fValue[i+1] - fValue[i-1])/(maxFval - minFval)
```

and not the average of crowdingDistances

Reply:
I've implemented it in this  [patch](https://drive.google.com/file/d/1KjrbkkKL6eviF7dUEjsCNr8BCW-6c4rg/view)

Reply:
Good, can you make the patch part of the PR?

Reply:
I've no problem with that. Thing is, @say4n  has implemented assignment first and sorting later which I think is wrong but I guess he must have some reason. I was waiting for his review, very well, I'll add this patch on my PR.

Reply:
I think I had unrolled the first iteration of some parts of the optimisation loop to be outside the for loop. This had the benefit of instantiating some items as far as I recall. I don't remember all the details now so do take this with a grain of salt! :)

Reply:
Thanks for clarifying :). I'd say let's merge my patch for now as it is straightforward & adheres to the research paper directly. We can then try your optimization method and see how it all fits in.

Reply:
Yes, for sure! Great work btw!!